Add demo for table recognition in youtube video

maximumSHOT-HSE opened the pull request.

maximumSHOT-HSE committed:
Add demo for table recognition in youtube video

goodwinnk commented on table_recognition/demo/local_data_demo.sh:
```diff
@@ -2,15 +2,12 @@ export PYTHONPATH=$PYTHONPATH:../..
 VIDEO_PATH="../../data/sync/table_recognition_demo/table_recongition_demo_video.mp4"
 LAYOUT_PATH="layout.txt"
-VIDEO_WITH_TABLE_PATH=video.mp4
+VIDEO_WITH_TABLE_PATH=local_video_table.mp4
```
I think it's ok to use python scripts for demonstration purposes, at least it will solve problems with running on Windows.

Here's the batch file:

```
@echo off

set PYTHONPATH=%PYTHONPATH%;../..

set VIDEO_PATH="../../data/sync/table_recognition_demo/table_recongition_demo_video.mp4"
set LAYOUT_PATH="layout.txt"
set VIDEO_WITH_TABLE_PATH=local_video_table.mp4

echo "Finding frame by frame table polygon..."
python ../find_table_polygon.py --video %VIDEO_PATH% --layout %LAYOUT_PATH%

echo "Drawing table..."
python ../highlight_table.py --video %VIDEO_PATH% --layout %LAYOUT_PATH% --output %VIDEO_WITH_TABLE_PATH%

echo "Video with highlighted table saved in: %VIDEO_WITH_TABLE_PATH%"
```

goodwinnk commented on table_recognition/demo/youtube_data_demo.sh:
```diff
@@ -0,0 +1,24 @@
+export PYTHONPATH=$PYTHONPATH:../..
+
+VIDEO_FOLDER="youtube_folder"
+VIDEO_PATH="$VIDEO_FOLDER/youtube_video.mp4"
```
Please use `repo/data/local` directory for the demonstration purposes. This directory is ignored, so writing to it won't leave dirty files in the repository.

goodwinnk commented on data_utils/data_cli.py:
```diff
@@ -46,5 +68,7 @@ def run_manual_extraction(args):
                                        help='Output directory. Directory of video is used by default.')
     extract_manual_parser.set_defaults(func=run_manual_extraction)
 
+    add_cut_subparser(subparsers)
```
Хорошо, что выделил функцию! Плохо, что теперь в репозитории два стиля задания команд. Когда это просто, то лучше рефакторить по ходу дела.

goodwinnk commented on data_utils/data_cli.py:
```diff
@@ -15,6 +18,25 @@ def run_manual_extraction(args):
                         frame_save_modifier=args.name_mixture)
 
 
+def run_cut(args):
```
Раз теперь выделилась функция, то `run_cut `можно было бы делать локальной в `add_cut_subparser`.

goodwinnk commented on data_utils/utils.py:
```diff
@@ -146,6 +147,37 @@ def to_gray_gaussian(frame, index):
+        if not response or frame_id > to_frame:
+            break
+        print(frame_id)
+        if from_frame <= frame_id:
```
Думаю, стоило переставить видео на `from_frame` через `video.set(cv2.CAP_PROP_POS_FRAMES, frame_index)`.

goodwinnk commented on data_utils/utils.py:
```diff
@@ -146,6 +147,37 @@ def to_gray_gaussian(frame, index):
+        response, frame = capture.read()
+        if not response or frame_id > to_frame:
+            break
+        print(frame_id)
```
Писать в общий поток из утилитных функций - плохая идея. Потом намучаемся с этим, но пока можно ничего не делать.

maximumSHOT-HSE replied:
Забыл удалить лишний debug-вывод.

goodwinnk commented on table_recognition/demo/clear.sh:
```diff
@@ -1,2 +1,4 @@
+rm -rf youtube_folder
```
Можно чистить перед запуском, чтобы не приходилось об этом думать при старте демки, чтобы не было необходимости знать про этот дополнительный файл и чтобы не приходилось дублировать константы в нескольких местах.

maximumSHOT-HSE committed:
Refactor subparsers creating process

maximumSHOT-HSE committed:
Remove debug printings

maximumSHOT-HSE committed:
Wrtie demo using python instead of bash

maximumSHOT-HSE committed:
Fix small bug realted with PYTHONPATH env variable settings

maximumSHOT-HSE committed:
Change start frame index processing in video cutting

goodwinnk commented on table_recognition/demo/clear.sh:
```diff
@@ -1,4 +0,0 @@
-rm -rf youtube_folder
```
Нерелевантные изменения оказались в коммите.

goodwinnk commented on table_recognition/demo/run_demo.py:
```diff
@@ -0,0 +1,77 @@
+if __name__ == '__main__':
+
+    if os.path.exists(GENERATED_PATH):
+        shutil.rmtree(GENERATED_PATH)
```
👍

goodwinnk commented on table_recognition/demo/run_demo.py:
```diff
@@ -0,0 +1,77 @@
+def draw_table(input_video_path, layout_path, video_with_table_path):
+    print('Finding frame by frame table polygon...')
+
+    subprocess.run([
```
Зачем нужен `subprocess`? Как раз демо помогает понять, как код предстоит использовать и создает дополинтельную разбивку на функции. Плюс отладка, я так понимаю, тоже на этом заткнется. 

Давай выложу эти два коммита: https://github.com/goodwinnk/billiards/compare/master...048ae517fba05085e9dcd97906a272f0c5622ff1.

maximumSHOT-HSE replied:
Демка с youtube-видео пропала, я перепишу без subprocess.

maximumSHOT-HSE committed:
Move hh:mm:ss parsing function from cli to utils lib

maximumSHOT-HSE committed:
Rewrite demo script without using subprocess script

Demo script should help users for better code unserstanding and
fast results retrieving. For this it is better to get rid of
subprocess and use functionality as library functions
